torch/utils/_config_module.py

```diff
@@ -444,11 +443,12 @@ def _is_default(self, name: str) -> bool:
             config_val.env_value_force is _UNSET_SENTINEL
             or config_val.env_value_force == config_val.default
         )
-        return (
-            config_val.user_override is _UNSET_SENTINEL
-            and not_set_env_default
-            and not_set_env_force
-        )
+
+        unset = config_val.user_override is _UNSET_SENTINEL
+        # Handle reference types specially to avoid spammy warnings
+        if isinstance(config_val.default, (list, set, dict)):
+            unset = unset or config_val.user_override == config_val.default
```

Comment:
is there a reason why we wouldn't default to value equality more generally ?

Reply:
We can, for all types originally just comparing with _UNSET_SENTINEL was slightly better., but then we had to cludge around with reference semantics.

The one trick is we only allow a lot of overrides if the value is not set by the user (which breaks for reference types, but they also don't allow overrides anyway).